File: spring-web-java/src/main/java/spring/web/java/presentation/exception/handler/ExceptionHandlerFilter.java
package spring.web.java.presentation.exception.handler;

import java.io.IOException;

import org.springframework.http.HttpStatus;
import org.springframework.http.ProblemDetail;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import io.jsonwebtoken.JwtException;
import jakarta.servlet.FilterChain;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import spring.web.java.infrastructure.logging.HttpLogFilter;
import spring.web.java.infrastructure.util.ResponseWriter;

@Slf4j
@Component
@RequiredArgsConstructor
public class ExceptionHandlerFilter extends OncePerRequestFilter {

	private final ResponseWriter responseWriter;

	@Override
	protected void doFilterInternal(
		HttpServletRequest request,
		HttpServletResponse response,
		FilterChain filterChain
	) throws IOException {
		try {
			filterChain.doFilter(request, response);
		} catch (JwtException e) {
			handleException(HttpStatus.UNAUTHORIZED, e.getMessage());
		} catch (Exception e) {
			log.error("[{}]", request.getAttribute(HttpLogFilter.TRANSACTION_ID), e);
			handleException(HttpStatus.INTERNAL_SERVER_ERROR, e.getMessage());
		}
	}

	private void handleException(HttpStatus status, String message) throws IOException {
		responseWriter.writeResponse(ProblemDetail.forStatusAndDetail(status, message));
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/presentation/exception/handler/ExceptionHandlerFilter.kt
package spring.web.kotlin.presentation.exception.handler

import io.jsonwebtoken.JwtException
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpStatus
import org.springframework.http.ProblemDetail
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import spring.web.kotlin.infrastructure.util.ResponseWriter

@Component
class ExceptionHandlerFilter(
    private val responseWriter: ResponseWriter
) : OncePerRequestFilter() {
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        runCatching { filterChain.doFilter(request, response) }
            .onFailure { throwable ->
                when (throwable) {
                    is JwtException -> handleException(HttpStatus.UNAUTHORIZED, throwable.message)

                    is Exception -> handleException(HttpStatus.INTERNAL_SERVER_ERROR, throwable.message)
                }
            }
    }

    private fun handleException(status: HttpStatus, message: String?) {
        val problemDetail = ProblemDetail.forStatus(status).apply { detail = message }
        responseWriter.writeResponse(problemDetail)
    }
}
